crates/ide_assists/src/handlers/convert_tuple_struct_to_named_struct.rs

```diff
@@ -0,0 +1,431 @@
+    let usages = strukt_def.usages(&ctx.sema).include_self_kw_refs(true).all();
+
+    let edit_node = |edit: &mut AssistBuilder, node: SyntaxNode| -> Option<()> {
+        match_ast! {
+            match node {
+                ast::TupleStructPat(tuple_struct_pat) => {
+                    edit.replace(
+                        tuple_struct_pat.syntax().text_range(),
+                        ast::make::record_pat_with_fields(
+                            tuple_struct_pat.path()?,
+                            ast::make::record_pat_field_list(tuple_struct_pat.fields().zip(names).map(
+                                |(pat, name)| {
+                                    ast::make::record_pat_field(
+                                        ast::make::name_ref(&name.to_string()),
+                                        pat,
+                                    )
+                                },
+                            )),
+                        )
+                        .to_string(),
+                    );
+                },
+                // for tuple struct creations like Foo(42)
+                ast::CallExpr(call_expr) => {
+                    let path = call_expr.syntax().descendants().find_map(ast::PathExpr::cast).and_then(|expr| expr.path())?;
+
+                    // this also includes method calls like Foo::new(42), we should skip them
+                    if let Some(name_ref) = path.segment().and_then(|s| s.name_ref()) {
+                        match NameRefClass::classify(&ctx.sema, &name_ref) {
+                            Some(NameRefClass::Definition(Definition::SelfType(_))) => {},
+                            Some(NameRefClass::Definition(def)) if def == strukt_def => {},
+                            _ => return None,
+                        };
+                    }
+
+                    let arg_list = call_expr.syntax().descendants().find_map(ast::ArgList::cast)?;
+
+                    edit.replace(
+                        call_expr.syntax().text_range(),
+                        ast::make::record_expr(
+                            path,
+                            ast::make::record_expr_field_list(arg_list.args().zip(names).map(
+                                |(expr, name)| {
+                                    ast::make::record_expr_field(
+                                        ast::make::name_ref(&name.to_string()),
+                                        Some(expr),
+                                    )
+                                },
+                            )),
+                        )
+                        .to_string(),
+                    );
+                },
+                _ => ()
+            }
+        }
+        Some(())
+    };
+
+    for (file_id, refs) in usages {
+        edit.edit_file(file_id);
+        for r in refs {
+            for node in r.name.syntax().ancestors() {
+                edit_node(edit, node);
```

Comment:
If I'm not mistaken the code here would try to replace both patterns in `SomeTupleStruct(TupleStructWeWantToReplace(...))` for example, so we should check the path of the pattern here as well. We can also exit the loop once we replaced one thing since one ref should only trigger one replacement.

Otherwise I think this is good to go
bors d+

Reply:
Exiting loop once we replaced one thing makes nested tuple [test](https://github.com/rust-analyzer/rust-analyzer/pull/8317/commits/97270dfb9162bfcefa78a7ed39551069ae267f48#diff-8eb30214a5bdac1d5eb43ccdc3eddd3c669f4ec2f38295aa2f01edc163f58c7eR382-R424) pass, should i add a check for the path anyway?

Reply:
Yes otherwise the inverted nesting could fail, as in `TupleStructWeWantToReplace(SomeTupleStruct(...))` will replace the inner one but it should replace the outer one.

Reply:
the reverse [test](https://github.com/rust-analyzer/rust-analyzer/pull/8317/commits/5e765895cf6fea42649a846c0297d95a9aa7b162#diff-8eb30214a5bdac1d5eb43ccdc3eddd3c669f4ec2f38295aa2f01edc163f58c7eR424-R460) case is also passing, i guess [this check](https://github.com/rust-analyzer/rust-analyzer/pull/8317/files#diff-8eb30214a5bdac1d5eb43ccdc3eddd3c669f4ec2f38295aa2f01edc163f58c7eR140) does the trick. or am i missing something?

Reply:
Ah no, that check is irrelevant to the pattern stuff, but we are using the direct ancestors of the nameref so we don't even see the underlying patterns I didn't realize that so ye this is fine as is I believe

bors r+